R/bs-theme.R

```diff
@@ -1,3 +1,226 @@
+#' @param secondary A color for components and messages that don't need to stand
+#'   out. (Not supported in Bootstrap 3.)
+#' @param success A color for messages that indicate an operation has succeeded.
+#'   Typically green.
+#' @param info A color for messages that are informative but not critical.
+#'   Typically a shade of blue-green.
+#' @param warning A color for warning messages. Typically yellow.
+#' @param danger A color for errors. Typically red.
+#' @param base_font The default typeface.
+#' @param code_font The typeface to be used for code. Be sure this is monospace!
+#' @param heading_font The typeface to be used for heading elements.
+#'
+#' @references \url{https://getbootstrap.com/docs/4.4/getting-started/theming/}
+#' @references \url{https://rstudio.github.io/sass/}
+#' @seealso [bs_add_variables()], [bs_theme_preview()], [bs_dependencies()],
+#'   [bs_global_set()]
+#' @examples
+#'
+#' theme <- bs_theme(
+#'   # Controls the default grayscale palette
+#'   bg = "#202123", fg = "#B8BCC2",
+#'   # Controls the accent (e.g., hyperlink, button, etc) colors
+#'   primary = "#EA80FC", secondary = "#48DAC6",
+#'   base_font = c("Grandstander", "sans-serif"),
+#'   code_font = c("Courier", "monospace"),
+#'   heading_font = "'Helvetica Neue', Helvetica, sans-serif",
+#'   # Can also add lower-level customization
+#'   "input-border-color" = "#EA80FC"
+#' )
+#' if (interactive()) {
+#'   bs_theme_preview(theme)
+#' }
+#'
+#' # Lower-level bs_add_*() functions allow you to work more
+#' # directly with the underlying Sass code
+#' theme <- theme %>%
+#'   bs_add_variables("my-class-color" = "red") %>%
+#'   bs_add_rules(".my-class { color: $my-class-color }")
+#'
+#' @export
+bs_theme <- function(version = version_default(), bootswatch = NULL, ...,
+                     bg = NULL, fg = NULL, primary = NULL, secondary = NULL,
+                     success = NULL, info = NULL, warning = NULL, danger = NULL,
+                     base_font = NULL, code_font = NULL, heading_font = NULL) {
+  version <- version_resolve(version)
+  bootswatch <- bootswatch_theme_resolve(bootswatch, version)
+  theme <- bs_add_layers(
+    bs_theme_init(),
+    bootstrap_layer(version),
+    if (identical(version, "4+3")) bs3compat_layer(),
+    # This will set a $navbar-height Sass var, even if no bootswatch is used
+    # TODO: maybe make navbar adjustment via jQuery instead? https://stackoverflow.com/a/46021836/1583084
+    navbar_height_layer(bootswatch, version),
+    bootswatch_layer(bootswatch, version)
+  )
+  bs_theme_update(
+    theme, ...,
+    bg = bg, fg = fg,
+    primary = primary,
+    secondary = secondary,
+    success = success,
+    info = info,
+    warning = warning,
+    danger = danger,
+    base_font = base_font,
+    code_font = code_font,
+    heading_font = heading_font
+  )
+}
+
+#' @rdname bs_theme
+#' @param theme a [bs_theme()] object.
+#' @export
+bs_theme_update <- function(theme = bs_theme(), ..., bg = NULL, fg = NULL,
```

Comment:
Do we need a default value of `theme = bs_theme()`? Presumably if people call `bs_theme_update`, they have a thing they want to update.

Reply:
No, we don't -- I was thinking it might make for useful short-hand code (i.e., `bs_theme_update(primary = "red")`), but I'm maybe we should make it more explicit

Reply:
If they call `bs_theme_update(primary = "red")` and intend it to create a new theme to modify, they might as well call `bs_theme(primary = "red")`, right?

I think it would be a little surprising to create a new theme object automatically for `bs_theme_create`(). So I'd vote for removing the default value for `theme`.

Reply:
Yea, agreed